File: src/AdvanceGameState.hs
module AdvanceGameState where

checa :: [[Char]] -> Int -> Int -> Bool
checa mx y x =
  if x == 0 || y == 0 || x == length mx -1 || y == length (mx!!0) -1
  then False
  else if mx!!x!!y == '#'
  then (liveNeighbours >= 2 && liveNeighbours <= 3)
  else (liveNeighbours == 3)

  where 
    aux = [mx!!(x-1)!!y, mx!!x!!(y-1), mx!!(x+1)!!y, mx!!x!!(y+1), mx!!(x+1)!!(y+1), mx!!(x-1)!!(y+1), mx!!(x+1)!!(y-1), mx!!(x-1)!!(y-1)]
    liveNeighbours = sum [1 | x<-aux, x == '#']

advanceGameState :: [[Char]] -> [[Char]]
advanceGameState mx =
  [[if checa mx x y then '#' else ' ' | x <- [0..(w-1)]]| y <- [0..(h-1)]]
  where
    w = length (mx!!1)
    h = length mx

File: prolog/AdvanceGameState.pl




% Conta # funcional
conte([], 0).
conte(['#'|T], N) :- conte(T,N1), N is N1 + 1.
conte([X|T], N) :- X \= '#', conte(T,N). % talvez seja redundante a primiera condicional




checa(Mx, Y, X) :-
    %aux([A, B, C, D, E, F, G, H]),
    nth0(0, Mx, List),
    length(List, Row), 
    Len is Row - 1, 
    
	(X =\= 0, Y =\= 0, length(Mx, K), K =\= X, Y =\= Len),
    
    Xx is X -1, XX is X + 1, Yy is Y - 1, YY is Y + 1,
    
    nth0(Y, Aa, A), nth0(Xx, Mx, Aa),
    nth0(Yy, Bb, B), nth0(X, Mx, Bb),
    nth0(Y, Cc, C), nth0(XX, Mx, Cc),
    nth0(Yy, Dd, D), nth0(X, Mx, Dd),
    nth0(YY, Ee, E), nth0(XX, Mx, Ee),
    nth0(YY, Ff, F), nth0(Xx, Mx, Ff),
    nth0(Yy, Gg, G), nth0(XX, Mx, Gg),
    nth0(Yy, Hh, H), nth0(Xx, Mx, Hh),

    %aux(Aux),
    conte([A, B, C, D, E, F, G, H], LiveNeighbours),
    
    (nth0(Y, Bb, I), I = '#' -> LiveNeighbours >= 2, LiveNeighbours =< 3).



advanceGameState(Mx, _, Y, []) :- length(Mx, Y). % Se manter o método como recebendo dois parâmetros, refatorar todas as suas ocorrências
advanceGameState(Mx, X, Y, [H2|T2]) :- % X e Y começam com 0 (zero).
    advanceGameStateRow(Mx, X, Y, H2), 
    K is Y + 1,
    advanceGameState(Mx, 0, K, T2).
    
advanceGameStateRow(Mx, X, _, []) :- nth0(0, Mx, R), length(R, X).
advanceGameStateRow(Mx, X, Y, [H2|T2]) :- 
    (checa(Mx, X, Y) -> H2 = '#'; H2 = ' '),
    Z is X + 1,
    advanceGameStateRow(Mx, Z, Y, T2).
     